src/interactions/panZoomInteraction.ts

```diff
@@ -20,56 +22,59 @@ export module Interaction {
-      }
-      if (yScale) {
-        this._yScale = yScale;
-        // HACKHACK #1388: self-register for resetZoom()
-        this._yScale.broadcaster.registerListener("pziY" + this.getID(), () => this.resetZoom());
-      }
+      this._xScale = xScale;
+      this._yScale = yScale;
+
+      this._dragInteraction = new Interaction.Drag();
+      this._scrollInteraction = new Interaction.Scroll();
+      this._setupInteractions();
     }
 
-    /**
-     * Sets the scales back to their original domains.
-     */
-    public resetZoom() {
-      // HACKHACK #254
-      this._zoom = d3.behavior.zoom();
-      if (this._xScale) {
-        this._zoom.x((<any> this._xScale)._d3Scale);
-      }
-      if (this._yScale) {
-        this._zoom.y((<any> this._yScale)._d3Scale);
-      }
-      this._zoom.on("zoom", () => this._rerenderZoomed());
-      this._zoom(this._hitBox);
+    public _requiresHitbox() {
+      return true;
     }
 
     public _anchor(component: Component.AbstractComponent, hitBox: D3.Selection) {
       super._anchor(component, hitBox);
-      this.resetZoom();
+      this._dragInteraction._anchor(component, hitBox);
+      this._scrollInteraction._anchor(component, hitBox);
     }
 
-    public _requiresHitbox() {
-      return true;
+    private _setupInteractions() {
+      this._setupDragInteraction();
+      this._setupScrollInteraction();
     }
 
-    private _rerenderZoomed() {
-      // HACKHACK since the d3.zoom.x modifies d3 scales and not our TS scales, and the TS scales have the
-      // event listener machinery, let's grab the domain out of the d3 scale and pipe it back into the TS scale
-      if (this._xScale) {
-        var xDomain = (<any> this._xScale)._d3Scale.domain();
-        this._xScale.domain(xDomain);
-      }
+    private _setupDragInteraction() {
+      var lastDragPoint: Point;
+      this._dragInteraction.drag((startPoint, endPoint) => {
+        var dragAmountX = endPoint.x - (lastDragPoint == null ? startPoint.x : lastDragPoint.x);
+        var dragAmountY = endPoint.y - (lastDragPoint == null ? startPoint.y : lastDragPoint.y);
+        if (this._xScale != null) {
+          this._xScale.domain(ScaleDomainTransformers.translate(this._xScale, -dragAmountX));
+        }
+        if (this._yScale != null) {
+          this._yScale.domain(ScaleDomainTransformers.translate(this._yScale, -dragAmountY));
+        }
+        lastDragPoint = endPoint;
+      });
+      this._dragInteraction.dragend(() => lastDragPoint = null);
+    }
 
-      if (this._yScale) {
-        var yDomain = (<any> this._yScale)._d3Scale.domain();
-        this._yScale.domain(yDomain);
-      }
+    private _setupScrollInteraction() {
+      var magnifyAmount = 1;
+      this._scrollInteraction.onScroll((point: Point, deltaAmount: number) => {
+        var oldMagnifyAmount = magnifyAmount;
+        magnifyAmount = Math.pow(2, -deltaAmount * .002) * magnifyAmount;
```

Comment:
This code doesn't seem to be very clear -- the interaction between `oldMagnifyAmount` and `newMagnifyAmount` in particular is kind of confusing. It seems like we should be able to calculate the ratio directly, without storing the value off twice...?